### PR TITLE
Scala dist training

### PR DESCRIPTION
**Pull request by javelinjs**
Make scala package support distributed training. the very initial step for https://github.com/dmlc/mxnet/issues/1637

I am going to make it able to initialize PS by passing configs from function arguments instead of env variables in the following days.

### CODE REVIEW

**Comment by javelinjs** [comment]
@tqchen Mind helping to double check the change made for `MXKVStoreRunServer` in c api?

**Comment by piiswrong** [comment]
when you change c_api it's usually better to make a new one, like MXKVStoreRunServerEx. Otherwise it will break all the other language interfaces.

File: include/mxnet/c_api.h [hunk 1]
```diff
@@ -1153,19 +1153,23 @@ MXNET_DLL int MXKVStoreBarrier(KVStoreHandle handle);
  * \brief the prototype of a server controller
  * \param head the head of the command
  * \param body the body of the command
+ * \param controller_handle helper handle for implementing controller
  */
 typedef void (MXKVStoreServerController)(int head,
-                                         const char* body);
+                                         const char *body,
+                                         void *controller_handle);
```
Comment (mli) [inline include/mxnet/c_api.h]: is that necessary? another way may be pack the handle into the controller. for example, in c++

```
controller = [handle](int head, const char* body) { handle->xxx(); } 
```

does a similar solution work for you?

**Comment by tqchen** [comment]
I agree that explicitly having handle is always better. As closure should not be treated as function. 

@mli As long as we make sure the old calls to server works, I think this change is fine.

**Comment by javelinjs** [comment]
Yah, the explicit handle makes front-end rely on pure C api.

BTW, since @piiswrong suggested to make it a new name, I haven't update this PR. I'm trying to figure out what is the most natural way to run the KVStoreServer, which is related to arguments this function takes, according to https://github.com/dmlc/mxnet/issues/1637

**Comment by mli** [comment]
keeping the same name should be fine.

**Comment by javelinjs** [comment]
OK, this time let's keep the same name.

**Pull request merged**